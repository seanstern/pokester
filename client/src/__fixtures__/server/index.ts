export { default, rest } from "./server";
export { validRoomIdForPatch } from "./handlers";
