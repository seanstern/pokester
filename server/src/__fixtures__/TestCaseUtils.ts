import { Fixture, FixtureModule } from "./TypeUtils";

/**
 * Given a fixture module, returns each fixture as cases table that
 * can be consumed by jest's describe.each
 * (c.f. https://jestjs.io/docs/api#describeeachtablename-fn-timeout)
 * or test.each
 * (c.f https://jestjs.io/docs/api#testeachtablename-fn-timeout)
 *
 * @example
 * import * as FooFixtureModule from "Foo.fixture";
 * const fooCasesTable = toJestCasesTable(FooFixtureModule);
 * describe.each(fooCasesTable)("$description", ({ create }) => {
 *   const fixture = create();
 *   ...
 * })
 *
 * @param fixtureModule a fixture module
 * @returns a fixture cases table that can be consumed by jest's
 *   describe.each or test.each
 */
export const asJestCasesTable = <T>(
  fixtureModule: FixtureModule<T>
): Fixture<T>[] => Object.values(fixtureModule);
