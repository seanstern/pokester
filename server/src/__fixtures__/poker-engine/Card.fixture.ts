import { Card, CardSuit, CardRank } from "@chevtek/poker-engine";
import { FixtureModule } from "../TypeUtils";

type CardSuitEntry = [string, CardSuit];
const cardSuits: CardSuitEntry[] = Object.entries(CardSuit);
type CardRankEntry = [string, CardRank];
const cardRanks: CardRankEntry[] = Object.entries(CardRank);
const allCardRankSuitEntryTuples = cardSuits.flatMap((cardSuitEntry) =>
  cardRanks.map(
    (cardRankEntry) =>
      [cardRankEntry, cardSuitEntry] as [CardRankEntry, CardSuitEntry]
  )
);

const cardFixtureModule: FixtureModule<Card> = allCardRankSuitEntryTuples.reduce(
  (cardsFixtureModule, [cardRankEntry, cardSuitEntry]) => {
    const [cardRankName, cardRankValue] = cardRankEntry;
    const camelCaseCardRankName = cardRankName.toLowerCase();
    const [cardSuitName, cardSuitValue] = cardSuitEntry;
    const camelCaseCardSuitName = `${cardSuitName
      .substring(0, 1)
      .toUpperCase()}${cardSuitName.substring(1).toLowerCase()}s`;
    const fixtureName = `${camelCaseCardRankName}Of${camelCaseCardSuitName}`;
    cardsFixtureModule[fixtureName] = {
      description: `${cardRankName.toLowerCase()} of ${cardSuitName.toLowerCase()}s`,
      create: () => new Card(cardRankValue, cardSuitValue),
    };
    return cardsFixtureModule;
  },
  {} as FixtureModule<Card>
);

export default cardFixtureModule;
