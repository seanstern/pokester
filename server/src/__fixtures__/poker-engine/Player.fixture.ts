import { Player } from "@chevtek/poker-engine";
import { Fixture } from "../TypeUtils";
import {
  onePlayerSeated,
  playersSeated,
  preFlop,
  completeRound,
  flop,
  turn,
} from "./Table.fixture";

export const onlyDealerSeated: Fixture<Player> = {
  description: "Dealer seated alone at table, pre-game",
  create: () => {
    const player = onePlayerSeated.create().dealer;
    if (!player) {
      throw new Error("No dealer in onePlayerSeated Table.fixture");
    }
    return player;
  },
};

export const dealerAmongSeatedPlayers: Fixture<Player> = {
  description: "Dealer seated at table with players, pre-game",
  create: () => {
    const player = playersSeated.create().dealer;
    if (!player) {
      throw new Error("No dealer in playersSeated Table.fixture");
    }
    return player;
  },
};

export const smallBlindAmongSeatedPlayers: Fixture<Player> = {
  description: "Small blind player seated at table with players, pre-game",
  create: () => {
    const player = playersSeated.create().smallBlindPlayer;
    if (!player) {
      throw new Error("No smallBlindPlayer in playersSeated Table.fixture");
    }
    return player;
  },
};

export const dealerAtFlop: Fixture<Player> = {
  description: "Dealer at flop",
  create: () => {
    const player = flop.create().dealer;
    if (!player) {
      throw new Error("No dealer in playersSeated Table.fixture");
    }
    return player;
  },
};

export const priorActorPostDealPreFlop: Fixture<Player> = {
  description:
    "Player immediately priort to current actor, immeidately after cards dealt",
  create: () => {
    const { lastActor } = preFlop.create();
    if (!lastActor) {
      throw new Error("No lastActor in preFlop Table.fixture");
    }
    return lastActor;
  },
};

export const currentActorPostDealPreFlop: Fixture<Player> = {
  description: "Player who is current actor, immediately after cards dealt",
  create: () => {
    const { currentActor } = preFlop.create();
    if (!currentActor) {
      throw new Error("No currentActor in preFlop Table.fixture");
    }
    return currentActor;
  },
};

export const folded: Fixture<Player> = {
  description:
    "Player who has folded during flop, immediately at start of turn",
  create: () => {
    const { players } = turn.create();
    const foldedPlayer = players.find((player) => player?.folded === true);
    if (!foldedPlayer) {
      throw new Error("No foldedPlayer in turn Table.fixture");
    }
    return foldedPlayer;
  },
};

export const dealerStoodUpAtEndOfRound: Fixture<Player> = {
  description: "Dealer who has stood up, immediately after completion of round",
  create: () => {
    const { dealer } = completeRound.create();
    if (!dealer) {
      throw new Error("No dealer in completeRound Table.fixture");
    }
    if (!dealer.left) {
      throw new Error("Dealer has not stood up in completeRound Table.fixtre");
    }
    return dealer;
  },
};

export const nonDealerPlayerWithStoodUpDealerAtEndOfRound: Fixture<Player> = {
  description:
    "Non dealer at table with dealer who stood up, immediately after completion of round",
  create: () => {
    const table = completeRound.create();
    const nonDealer = table.players.find(
      (p) => p?.left === false && table.dealer !== p
    );
    if (!nonDealer) {
      throw new Error("No non dealer in completeRound Table.fixture");
    }
    return nonDealer;
  },
};

export const winner: Fixture<Player> = {
  description: "Player who is winner, immediately after completion of round",
  create: () => {
    const { winners } = completeRound.create();
    if (!winners) {
      throw new Error("No winners array in completeRound Table.fixture");
    }
    if (winners.length < 1) {
      throw new Error(
        "No elements in winners array in completeRound Table.fixture"
      );
    }
    return winners[0];
  },
};
