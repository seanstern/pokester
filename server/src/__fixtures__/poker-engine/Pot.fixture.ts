import { Pot } from "@chevtek/poker-engine";
import { Fixture } from "../TypeUtils";
import { preFlop, flop, completeRound } from "./Table.fixture";

/**
 * Given a potentially empty array of Pots, returns a Pot when the array
 * contains exactly one Pot; throws an error otherwise.
 * @param pots
 */
const getPot = (pots: Pot[]): Pot => {
  if (!pots) {
    throw new Error("No pots array in preFlop Table.fixture");
  }
  if (pots.length < 1) {
    throw new Error("No elements in pots array in preFlop Table.fixture");
  }
  return pots[0];
};

export const emptyPot: Fixture<Pot> = {
  description: "empty Pot",
  create: () => getPot(preFlop.create().pots),
};

export const potWithEligiblePlayersNoWinners: Fixture<Pot> = {
  description: "Pot with eligible players, no winners",
  create: () => {
    const pot = getPot(flop.create().pots);
    if (pot.eligiblePlayers.length < 1) {
      throw new Error(
        "No elements in eligiblePlayers array in flop Table.fixture"
      );
    }
    return pot;
  },
};

type WinnersPot = Pot & Required<Pick<Pot, "winners">>;
export const potWithWinners: Fixture<WinnersPot> = {
  description: "Pot with eligible players and winners",
  create: () => {
    const pot = getPot(completeRound.create().pots);
    if (!pot.winners) {
      throw new Error("No winners array in completeRound Table.fixture");
    }
    if (pot.winners.length < 1) {
      throw new Error(
        "No elements in winners array in completeRound Table.fixture"
      );
    }
    return pot as WinnersPot;
  },
};
