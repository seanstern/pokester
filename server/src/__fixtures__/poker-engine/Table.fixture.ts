import { Table, Card, CardSuit, CardRank } from "@chevtek/poker-engine";
import { Fixture } from "../TypeUtils";

/**
 * Returns a new deck with card ordered deterministically.
 *
 * Deterministic deck cration enables snapshot testing of
 * Table fixtures.
 *
 * @return a new deck with cards ordered deterministically
 */
const deterministicNewDeck: Table["newDeck"] = () => {
  const newDeck: Card[] = [];
  Object.keys(CardSuit).forEach((suit) => {
    Object.keys(CardRank).forEach((rank) => {
      newDeck.push(
        new Card(
          CardRank[rank as keyof typeof CardRank],
          CardSuit[suit as keyof typeof CardSuit]
        )
      );
    });
  });
  return newDeck;
};

export const emptyTable: Fixture<Table> = {
  description: "newly constructed empty Table",
  create: () => new Table(),
};

export const onePlayerSeated: Fixture<Table> = {
  description: "Table with one seated player",
  create: () => {
    const t = emptyTable.create();
    t.sitDown("John", 2000, 3);
    return t;
  },
};

export const playersSeated: Fixture<Table> = {
  description: "Table with seated players",
  create: () => {
    const t = onePlayerSeated.create();
    t.sitDown("Jane", 3000, 0);
    t.sitDown("Juan", 9199);
    t.sitDown("Jay", 31415, 7);
    return t;
  },
};

export const preFlop: Fixture<Table> = {
  description: "Table at start of pre-flop",
  create: () => {
    const t = playersSeated.create();
    // Shadow newDeck in object with deterministic version
    t.newDeck = deterministicNewDeck;
    t.dealCards();
    // Remove newDeck deterministic shadown from object *only*
    // newDeck continues to exist in Object.getPrototypeOf(t)
    delete (t as any).newDeck;
    return t;
  },
};

export const flop: Fixture<Table> = {
  description: "Table at start of flop; players have: called, checked",
  create: () => {
    const t = preFlop.create();
    if (!t.currentActor) {
      throw new Error(
        "t.currentActor in preFlop Table.fixture expected to be non-nullish"
      );
    }
    t.currentActor.callAction();
    t.currentActor.callAction();
    t.currentActor.callAction();
    t.currentActor.checkAction();
    return t;
  },
};

export const turn: Fixture<Table> = {
  description:
    "Table at start of turn; players have: called, checked, stoodUp, bet, folded",
  create: () => {
    const t = flop.create();
    if (!t.currentActor) {
      throw new Error(
        "t.currentActor in flop Table.fixture expected to be non-nullish"
      );
    }
    t.standUp("Jay");
    t.currentActor.betAction(50);
    t.currentActor.foldAction();
    t.currentActor.callAction();
    return t;
  },
};

export const river: Fixture<Table> = {
  description:
    "Table at start of river; players have called, checked, stoodUp, bet, folded, raised",
  create: () => {
    const t = turn.create();
    if (!t.currentActor) {
      throw new Error(
        "t.currentActor in turn Table.fixture expected to be non-nullish"
      );
    }
    t.currentActor.checkAction();
    t.currentActor.betAction(50);
    t.currentActor.raiseAction(100);
    t.currentActor.callAction();
    return t;
  },
};

export const completeRound: Fixture<Table> = {
  description:
    "Table at completion of round; players have called, checked, stoodUp, bet, folded, raised",
  create: () => {
    const t = river.create();
    if (!t.currentActor) {
      throw new Error(
        "t.currentActor in river Table.fixture expected to be non-nullish"
      );
    }
    t.currentActor.checkAction();
    t.currentActor.checkAction();
    return t;
  },
};

export const completeRoundWinnerStoodUp: Fixture<Table> = {
  description:
    "Table at completion of round; players have called, checked, stoodUp, bet, folded, raised; winner has stood up",
  create: () => {
    const t = completeRound.create();
    if (!t.winners || !t.winners[0]) {
      throw new Error(
        "t.winners in completeRound Table.fixture expeted to be non-nullish and have at least 1 element"
      );
    }
    t.standUp(t.winners[0]);
    return t;
  },
};

export const roundTwo: Fixture<Table> = {
  description: "Table at start of second round",
  create: () => {
    const t = completeRound.create();
    // Shadow newDeck in object with deterministic version
    t.newDeck = deterministicNewDeck;
    t.dealCards();
    // Remove newDeck deterministic shadown from object *only*
    // newDeck continues to exist in Object.getPrototypeOf(t)
    delete (t as any).newDeck;
    return t;
  },
};
