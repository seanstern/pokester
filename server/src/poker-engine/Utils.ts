import { Player } from "@chevtek/poker-engine";

/**
 * Given a Player, returns true when Player is Player.table.dealer
 * and Player.table.dealCards can be called without throwing an
 * exception; false otherwise.
 *
 * This code is meant to duplicate some of the internal validation
 * that Table.dealCards uses.
 *
 * @param p a Player
 * @returns true when player is Player.table.dealer and
 *   Player.table.dealCards can be called without throwing an
 *   exception, false otherwise
 */
export const canDealCards = (p: Player) => {
  const {
    table: { dealer, activePlayers, currentRound },
  } = p;
  return dealer === p && activePlayers.length >= 2 && !currentRound;
};
