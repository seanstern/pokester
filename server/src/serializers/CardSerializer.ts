import { Card, CardRank, CardSuit } from "@chevtek/poker-engine";
import {
  ArgumentsDeserializationSpec,
  createDeserializeArgumentsFn,
  Deserialize,
  JSONValue,
  serialize as serializeCommon,
} from "./CommonSerializer";

/**
 * Given an Card, returns a JSON conformant version of the Card
 * @param c a Card
 * @returns a JSON conformant version of the Card
 */
export const serialize: (c: Card) => JSONValue = serializeCommon;

/**
 * Given a JSONValue, returns true when it represent a CardRank
 * and false otherwise
 * @param json a JSONValue
 * @returns true when json is a CardRank, false otherwise
 */
const isCardRank = (json: JSONValue): json is CardRank =>
  typeof json === "string" && Object.values<string>(CardRank).includes(json);

/**
 * Given a JSONValue, returns true when it represent a CardSuit
 * and false otherwise
 * @param json a JSONValue
 * @returns true when json is a CardSuit, false otherwise
 */
const isCardSuit = (json: JSONValue): json is CardSuit =>
  typeof json === "string" && Object.values<string>(CardSuit).includes(json);

const constructorArgumentsDeserializationSpec: ArgumentsDeserializationSpec<
  ConstructorParameters<typeof Card>
> = [
  {
    serializedKeyName: "_rank",
    deserialize: (jsonValue) => {
      if (!isCardRank(jsonValue)) {
        throw new Error("Cannot deserialize JSON that is not CardRank");
      }
      return jsonValue;
    },
  },
  {
    serializedKeyName: "_suit",
    deserialize: (jsonValue) => {
      if (!isCardSuit(jsonValue)) {
        throw new Error("Cannot deserialize JSON that is not CardSuit");
      }
      return jsonValue;
    },
  },
];
const deserializeConstructorArgs = createDeserializeArgumentsFn(
  constructorArgumentsDeserializationSpec
);

/**
 * Given a JSON represenation of a Card, returns a Card.
 * @param json a JSON representation of a Card
 * @returns a Card
 */
export const deserialize: Deserialize<Card> = (json: JSONValue) =>
  new Card(...deserializeConstructorArgs(json));
