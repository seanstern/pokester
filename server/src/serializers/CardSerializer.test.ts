import { CardSuit, CardRank } from "@chevtek/poker-engine";
import CardFixtures from "../__fixtures__/poker-engine/Card.fixture";
import { asJestCasesTable } from "../__fixtures__/TestCaseUtils";
import { serialize, deserialize } from "./CardSerializer";

const cardCases = asJestCasesTable(CardFixtures);

describe("serialize produces valid JSON when given Card", () => {
  test.each(cardCases)("$description", ({ create }) => {
    const card = create();
    expect(serialize(card)).toMatchSnapshot();
  });
});

describe("deserialize", () => {
  describe("produces valid Card when given serialized version of Card", () => {
    test.each(cardCases)("$description", ({ create }) => {
      const card = create();
      const serializedCard = serialize(card);
      const deserializedCard = deserialize(serializedCard);
      expect(deserializedCard).not.toBe(card);
      expect(deserializedCard).toStrictEqual(card);
    });
  });

  describe("throws when given", () => {
    test("serialized Card with invalid rank", () => {
      expect(() =>
        deserialize({ _rank: "badrank", _suit: CardSuit.CLUB })
      ).toThrow(/CardRank[^]*_rank/);
    });

    test("serialized Card with invalid suit", () => {
      expect(() =>
        deserialize({ _rank: CardRank.EIGHT, _suit: "badsuit" })
      ).toThrow(/CardSuit[^]*_suit/);
    });
  });
});
