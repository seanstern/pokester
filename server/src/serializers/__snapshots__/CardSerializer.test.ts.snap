// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`serialize produces valid JSON when given Card ace of clubs 1`] = `
Object {
  "_rank": "A",
  "_suit": "c",
}
`;

exports[`serialize produces valid JSON when given Card ace of diamonds 1`] = `
Object {
  "_rank": "A",
  "_suit": "d",
}
`;

exports[`serialize produces valid JSON when given Card ace of hearts 1`] = `
Object {
  "_rank": "A",
  "_suit": "h",
}
`;

exports[`serialize produces valid JSON when given Card ace of spades 1`] = `
Object {
  "_rank": "A",
  "_suit": "s",
}
`;

exports[`serialize produces valid JSON when given Card eight of clubs 1`] = `
Object {
  "_rank": "8",
  "_suit": "c",
}
`;

exports[`serialize produces valid JSON when given Card eight of diamonds 1`] = `
Object {
  "_rank": "8",
  "_suit": "d",
}
`;

exports[`serialize produces valid JSON when given Card eight of hearts 1`] = `
Object {
  "_rank": "8",
  "_suit": "h",
}
`;

exports[`serialize produces valid JSON when given Card eight of spades 1`] = `
Object {
  "_rank": "8",
  "_suit": "s",
}
`;

exports[`serialize produces valid JSON when given Card five of clubs 1`] = `
Object {
  "_rank": "5",
  "_suit": "c",
}
`;

exports[`serialize produces valid JSON when given Card five of diamonds 1`] = `
Object {
  "_rank": "5",
  "_suit": "d",
}
`;

exports[`serialize produces valid JSON when given Card five of hearts 1`] = `
Object {
  "_rank": "5",
  "_suit": "h",
}
`;

exports[`serialize produces valid JSON when given Card five of spades 1`] = `
Object {
  "_rank": "5",
  "_suit": "s",
}
`;

exports[`serialize produces valid JSON when given Card four of clubs 1`] = `
Object {
  "_rank": "4",
  "_suit": "c",
}
`;

exports[`serialize produces valid JSON when given Card four of diamonds 1`] = `
Object {
  "_rank": "4",
  "_suit": "d",
}
`;

exports[`serialize produces valid JSON when given Card four of hearts 1`] = `
Object {
  "_rank": "4",
  "_suit": "h",
}
`;

exports[`serialize produces valid JSON when given Card four of spades 1`] = `
Object {
  "_rank": "4",
  "_suit": "s",
}
`;

exports[`serialize produces valid JSON when given Card jack of clubs 1`] = `
Object {
  "_rank": "J",
  "_suit": "c",
}
`;

exports[`serialize produces valid JSON when given Card jack of diamonds 1`] = `
Object {
  "_rank": "J",
  "_suit": "d",
}
`;

exports[`serialize produces valid JSON when given Card jack of hearts 1`] = `
Object {
  "_rank": "J",
  "_suit": "h",
}
`;

exports[`serialize produces valid JSON when given Card jack of spades 1`] = `
Object {
  "_rank": "J",
  "_suit": "s",
}
`;

exports[`serialize produces valid JSON when given Card king of clubs 1`] = `
Object {
  "_rank": "K",
  "_suit": "c",
}
`;

exports[`serialize produces valid JSON when given Card king of diamonds 1`] = `
Object {
  "_rank": "K",
  "_suit": "d",
}
`;

exports[`serialize produces valid JSON when given Card king of hearts 1`] = `
Object {
  "_rank": "K",
  "_suit": "h",
}
`;

exports[`serialize produces valid JSON when given Card king of spades 1`] = `
Object {
  "_rank": "K",
  "_suit": "s",
}
`;

exports[`serialize produces valid JSON when given Card nine of clubs 1`] = `
Object {
  "_rank": "9",
  "_suit": "c",
}
`;

exports[`serialize produces valid JSON when given Card nine of diamonds 1`] = `
Object {
  "_rank": "9",
  "_suit": "d",
}
`;

exports[`serialize produces valid JSON when given Card nine of hearts 1`] = `
Object {
  "_rank": "9",
  "_suit": "h",
}
`;

exports[`serialize produces valid JSON when given Card nine of spades 1`] = `
Object {
  "_rank": "9",
  "_suit": "s",
}
`;

exports[`serialize produces valid JSON when given Card queen of clubs 1`] = `
Object {
  "_rank": "Q",
  "_suit": "c",
}
`;

exports[`serialize produces valid JSON when given Card queen of diamonds 1`] = `
Object {
  "_rank": "Q",
  "_suit": "d",
}
`;

exports[`serialize produces valid JSON when given Card queen of hearts 1`] = `
Object {
  "_rank": "Q",
  "_suit": "h",
}
`;

exports[`serialize produces valid JSON when given Card queen of spades 1`] = `
Object {
  "_rank": "Q",
  "_suit": "s",
}
`;

exports[`serialize produces valid JSON when given Card seven of clubs 1`] = `
Object {
  "_rank": "7",
  "_suit": "c",
}
`;

exports[`serialize produces valid JSON when given Card seven of diamonds 1`] = `
Object {
  "_rank": "7",
  "_suit": "d",
}
`;

exports[`serialize produces valid JSON when given Card seven of hearts 1`] = `
Object {
  "_rank": "7",
  "_suit": "h",
}
`;

exports[`serialize produces valid JSON when given Card seven of spades 1`] = `
Object {
  "_rank": "7",
  "_suit": "s",
}
`;

exports[`serialize produces valid JSON when given Card six of clubs 1`] = `
Object {
  "_rank": "6",
  "_suit": "c",
}
`;

exports[`serialize produces valid JSON when given Card six of diamonds 1`] = `
Object {
  "_rank": "6",
  "_suit": "d",
}
`;

exports[`serialize produces valid JSON when given Card six of hearts 1`] = `
Object {
  "_rank": "6",
  "_suit": "h",
}
`;

exports[`serialize produces valid JSON when given Card six of spades 1`] = `
Object {
  "_rank": "6",
  "_suit": "s",
}
`;

exports[`serialize produces valid JSON when given Card ten of clubs 1`] = `
Object {
  "_rank": "T",
  "_suit": "c",
}
`;

exports[`serialize produces valid JSON when given Card ten of diamonds 1`] = `
Object {
  "_rank": "T",
  "_suit": "d",
}
`;

exports[`serialize produces valid JSON when given Card ten of hearts 1`] = `
Object {
  "_rank": "T",
  "_suit": "h",
}
`;

exports[`serialize produces valid JSON when given Card ten of spades 1`] = `
Object {
  "_rank": "T",
  "_suit": "s",
}
`;

exports[`serialize produces valid JSON when given Card three of clubs 1`] = `
Object {
  "_rank": "3",
  "_suit": "c",
}
`;

exports[`serialize produces valid JSON when given Card three of diamonds 1`] = `
Object {
  "_rank": "3",
  "_suit": "d",
}
`;

exports[`serialize produces valid JSON when given Card three of hearts 1`] = `
Object {
  "_rank": "3",
  "_suit": "h",
}
`;

exports[`serialize produces valid JSON when given Card three of spades 1`] = `
Object {
  "_rank": "3",
  "_suit": "s",
}
`;

exports[`serialize produces valid JSON when given Card two of clubs 1`] = `
Object {
  "_rank": "2",
  "_suit": "c",
}
`;

exports[`serialize produces valid JSON when given Card two of diamonds 1`] = `
Object {
  "_rank": "2",
  "_suit": "d",
}
`;

exports[`serialize produces valid JSON when given Card two of hearts 1`] = `
Object {
  "_rank": "2",
  "_suit": "h",
}
`;

exports[`serialize produces valid JSON when given Card two of spades 1`] = `
Object {
  "_rank": "2",
  "_suit": "s",
}
`;
